src/Microsoft.DocAsCode.Dfm/RendererCreator.cs

```diff
@@ -0,0 +1,182 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DocAsCode.Dfm
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Collections.Immutable;
+    using System.Linq;
+    using System.Reflection;
+    using System.Reflection.Emit;
+    using System.Threading;
+
+    using Microsoft.DocAsCode.MarkdownLite;
+    using Microsoft.DocAsCode.Common;
+
+    public static class RendererCreator
+    {
+        #region Fields
+        private static readonly ModuleBuilder _module = CreateModule();
+        private static readonly MethodInfo MatchMethod = typeof(IDfmRendererPart).GetMethod(nameof(IDfmRendererPart.Match));
+        private static readonly MethodInfo RenderMethod = typeof(IDfmRendererPart).GetMethod(nameof(IDfmRendererPart.Render));
+        private static readonly ConstructorInfo BaseConstructor = typeof(PlugableRendererBase).GetConstructor(new[] { typeof(object) });
+        private static readonly MethodInfo BaseRenderMethod = typeof(PlugableRendererBase).GetMethod(nameof(PlugableRendererBase.BaseRender));
+        private static int _typeCounter;
+        #endregion
+
+        #region Public Methods
+
+        public static object CreateRenderer(object baseRenderer, IEnumerable<IDfmRendererPartProvider> partProviders, IReadOnlyDictionary<string, object> parameters)
+        {
+            if (baseRenderer == null)
+            {
+                throw new ArgumentNullException(nameof(baseRenderer));
+            }
+            if (partProviders == null)
+            {
+                return baseRenderer;
+            }
+            return CreateRendererNoCheck(baseRenderer, partProviders, parameters ?? ImmutableDictionary<string, object>.Empty);
+        }
+
+        #endregion
+
+        #region Private Methods
+
+        private static ModuleBuilder CreateModule()
+        {
+            const string name = "dfm-renderer";
+            var dynamicAssembly = AppDomain.CurrentDomain.DefineDynamicAssembly(
+                new AssemblyName(name),
+                AssemblyBuilderAccess.Run);
+            return dynamicAssembly.DefineDynamicModule(name);
+        }
+
+        private static object CreateRendererNoCheck(object baseRenderer, IEnumerable<IDfmRendererPartProvider> partProviders, IReadOnlyDictionary<string, object> parameters)
```

Comment:
without check parameters.

---
In reply to: [110784577](https://github.com/dotnet/docfx/pull/1534#discussion_r110784577) [](ancestors = 110784577)